while-loops/problem-set-while-loops.md

```diff
@@ -0,0 +1,849 @@
+* title: While Loops
+
+##### !question
+
+Let's sum some numbers! We will write a function called `silly_sum` that reads a series of numbers from user input and returns the sum. To keep things interesting, let's also decide that entering `0` means the user is done entering input, and that we'll also stop asking for further input if the sum reaches or exceeds `1000`.
+
+Provide an implementation for `silly_sum` so that it
+- reads numbers from the user (use `input_int`) summing as we go until either
+    - the user enters `0`, or
+    - the sum reaches or exceeds `1000`
+
+##### !end-question
+
+##### !placeholder
+
+```python
+def silly_sum():
+    # Be sure to call input_int when you need an integer from the user.
+    # A default implementation is provided above for experimenting in VS Code or repl.it.
+    # You should not paste that implementation here. It will be available automatically.
+    pass
+```
+
+##### !end-placeholder
+
+##### !tests
+
+```python
+import unittest
+import main
+from main import silly_sum
+
+num_src = None
+
+def input_ints(nums):
+    for num in nums:
+        yield num
+
+def input_int(*args):
+    return next(num_src)
+
+# inject into main
+main.input_int = input_int
+
+class TestPython1(unittest.TestCase):
+
+    def test_one(self):
+        global num_src
+        num_src = input_ints([1, 1, 2, 3, 5, 8, 0, 9])
+        self.assertEqual(20, silly_sum(), 'the provided input was 1, 1, 2, 3, 5, 8, 0, 9')
+
+    def test_two(self):
+        global num_src
+        num_src = input_ints([400, 400, 400, 400])
+        self.assertEqual(1200, silly_sum(), 'the provided input was 400, 400, 400, 400')
+
+    def test_three(self):
+        global num_src
+        num_src = input_ints([999, 1, 1])
+        self.assertEqual(1000, silly_sum(), 'the provided input was 999, 1, 1')
+```
+
+##### !end-tests
+
+#### !hint
+
+A possible interaction with a user could look like this:
+> Enter an integer (0 to stop): 999<br />
+> Enter an integer (0 to stop): 1<br />
+> The sum is: 1000
+
+Here's another possible example:
+> Enter an integer (0 to stop): 400<br />
+> Enter an integer (0 to stop): 400<br />
+> Enter an integer (0 to stop): 400<br />
+> The sum is: 1200
+
+This sum is halted by the user:
+> Enter an integer (0 to stop): 1<br />
+> Enter an integer (0 to stop): 2<br />
+> Enter an integer (0 to stop): 3<br />
+> Enter an integer (0 to stop): 0<br />
+> The sum is: 6
+
+Note that printing the final sum line is not a listed requirement, but feel free to add additional output to your implementation to help think through the problem. It will not affect the results of the tests. Even if you _do_ print out the final sum, be sure to also _return_ it!
+
+#### !end-hint
+
+#### !explanation
+
+An example of a working implementation:
+
+```python
+def silly_sum():
+    # accumulate the sum
+    my_sum = 0
+
+    # stores the most recent user input
+    # default to anything except 0 (think about why this is)
+    num = 1
```

Comment:
I won't fight your implementation to change in Learn since it's ultimately a good conversation starter, but why not

```python
num = input_int()
```

?

Reply:
I don't like repeating the core part of the logic (fetching user input) if I can avoid it (especially as part of what amounts to just a variable declaration). Honestly, it kills me that Python doesn't have a do/while.

Personally, I wouldn't even use num in this implementation.  I'd just start up a while True and break as necessary!

I'm inclined to leave this as is.

Reply:
Makes sense to me! :)